File: deob/src/main/java/dev/openrs2/deob/classpath/AsmClassMetadata.java
package dev.openrs2.deob.classpath;

import java.util.List;
import java.util.stream.Collectors;

import dev.openrs2.asm.MemberDesc;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.ClassNode;

public final class AsmClassMetadata extends ClassMetadata {
	private final ClassPath classPath;
	private final ClassNode clazz;
	private final boolean dependency;

	public AsmClassMetadata(ClassPath classPath, ClassNode clazz, boolean dependency) {
		this.classPath = classPath;
		this.clazz = clazz;
		this.dependency = dependency;
	}

	@Override
	public String getName() {
		return clazz.name;
	}

	@Override
	public boolean isDependency() {
		return dependency;
	}

	@Override
	public boolean isInterface() {
		return (clazz.access & Opcodes.ACC_INTERFACE) != 0;
	}

	@Override
	public ClassMetadata getSuperClass() {
		if (clazz.superName != null) {
			return classPath.get(clazz.superName);
		}
		return null;
	}

	@Override
	public List<ClassMetadata> getSuperInterfaces() {
		return clazz.interfaces.stream()
			.map(classPath::get)
			.collect(Collectors.toUnmodifiableList());
	}

	@Override
	public List<MemberDesc> getFields() {
		return clazz.fields.stream()
			.map(f -> new MemberDesc(f.name, f.desc))
			.collect(Collectors.toUnmodifiableList());
	}

	@Override
	public List<MemberDesc> getMethods() {
		return clazz.methods.stream()
			.map(m -> new MemberDesc(m.name, m.desc))
			.collect(Collectors.toUnmodifiableList());
	}

	@Override
	public boolean isNative(MemberDesc method) {
		for (var m : clazz.methods) {
			if (m.name.equals(method.getName()) && m.desc.equals(method.getDesc())) {
				return (m.access & Opcodes.ACC_NATIVE) != 0;
			}
		}
		return false;
	}
}

File: asm/src/main/kotlin/org/openrs2/asm/classpath/AsmClassMetadata.kt
package org.openrs2.asm.classpath

import org.objectweb.asm.Opcodes
import org.objectweb.asm.tree.ClassNode
import org.openrs2.asm.MemberDesc

public class AsmClassMetadata(
    private val classPath: ClassPath,
    private val clazz: ClassNode,
    override val dependency: Boolean
) : ClassMetadata() {
    override val name: String
        get() = clazz.name

    override val `interface`: Boolean
        get() = clazz.access and Opcodes.ACC_INTERFACE != 0

    override val superClass: ClassMetadata?
        get() = clazz.superName?.let { classPath[it] ?: error("Failed to find $it on provided classpath.") }

    override val superInterfaces: List<ClassMetadata>
        get() = clazz.interfaces.map { classPath[it] ?: error("Failed to find $it on provided classpath.") }

    override val fields: List<MemberDesc>
        get() = clazz.fields.map(::MemberDesc)

    override val methods: List<MemberDesc>
        get() = clazz.methods.map(::MemberDesc)

    override fun getFieldAccess(field: MemberDesc): Int? {
        return clazz.fields.find { it.name == field.name && it.desc == field.desc }?.access
    }

    override fun getMethodAccess(method: MemberDesc): Int? {
        return clazz.methods.find { it.name == method.name && it.desc == method.desc }?.access
    }
}
